classes/userGroup/UserGroup.php

```diff
@@ -70,11 +70,11 @@ public function getContextId()
     /**
      * Set the context ID
      *
-     * @param int $contextId
+     * @param ?int $contextId
      */
     public function setContextId($contextId)
     {
-        $this->setData('contextId', $contextId);
+        $this->setData('contextId', (int) $contextId ?: null);
```

Comment:
I'd prefer if we didn't add extra casts etc. throughout the codebase (unless they are useful for backwards compatibility and explicitly flagged as such, for later removal). Casts like this aren't consistently used and they cause values to flip types unpredictably in places where you think you're just calling a getter/setter. It would be better if we adopted good first-class type hints throughout, and completely removed cases where there were special values, e.g. a special `0` (as noted above); then, if a `0` somehow snuck into the system (e.g. due to an out of date plugin or something), it would immediately trigger a referential integrity error. That would be quick to fix and not leave a trail of ambiguity behind it.